[1단계 - 로또(자동)] 유콩(김유빈) 미션 제출합니다.

kyukong opened the pull request:
안녕하세요 소니! 유콩입니다 😁
로또 미션 첫 PR 날립니다.

지난 미션에서 처음 접해보는 개념들이 많아
이번 미션에는 지난번에 알게 되었던 것들을 익숙해질 때까지 적용해보는 것이 목표입니다.
TDD, 일급 컬렉션, Stream, 인스턴스 변수 중 List 의 불변 등을 적용해보았습니다.
(우테코 페이지에 있는 1단계 피드백을 지금 단계에서 봐도 되는지 몰라서 적용을 못했습니다..!
리팩토링 시에 읽고 반영하겠습니다!!)

나름 열심히 해보았으나 코드에 찝찝함이 있어도 어떤 부분이 찝찝한지 찾기가 어렵네요. 🥲
피드백 기다리고 있겠습니다. ㅎㅎ

시간을 내주셔서 감사합니다! 🙇‍♀️

kyukong committed:
docs: 구현해야 하는 기능 작성

kyukong committed:
feat: 구입 금액 객체 생성

kyukong committed:
feat: 숫자 객체 생성

kyukong committed:
refactor(PurchaseAmount): 파라미터 변경 불가하도록 수정

kyukong committed:
feat: 로또 객체 생성

kyukong committed:
feat: 1에서 45 사이의 랜덤 값 가지는 Ball 객체 생성

kyukong committed:
feat: 입력받은 개수만큼 로또 객체들 생성

chawani committed:
fix(PurchaseAmount): 검증하는 로직에서 판단 오류 수정

chawani committed:
refactor: 코드 리팩터링

- 사용하지 않는 import 삭제
- 사용하지 않는 Ball 생성자 삭제
- Lotto 매직넘버 상수화
- 테스트코드 스네이크 케이스로 수정

chawani committed:
feat: 입력 금액에 맞는 로또 개수 반환 구현

chawani committed:
feat: 구입금액을 입력받고 로또들을 구매하는 기능 구현

chawani committed:
feat: 지난 주 당첨 번호 입력받는 기능 구현

chawani committed:
feat: 로또 결과 객체 생성

- Rank 열거형 생성

kyukong committed:
feat: 로또 일치 결과 계산하여 출력 및 Rank 클래스 속성 추가

- 간결한 네이밍 수정
- 매직넘버 상수화
- 테스트코드 간결화

kyukong committed:
feat: 수익률 계산하여 출력

kyukong committed:
refactor: 지난 주 당첨번호와 보너스 볼을 가지는 WinnigLotto 객체 생성

kyukong committed:
refactor: 입력 에러 발생 시 재입력 기능 추가

- Output 에서 에러 메시지 출력
- 사용하지 않는 util 클래스 삭제

kyukong committed:
fix: 숫자 판단 정규식 수정

- 음수도 포함

kyukong committed:
refactor: 코드 리팩터링

- Input, Output 메소드명 수정 및 호출 방식 수정
- 검증하는 메소드의 의도를 더 정확하게 나타내기 위해 validate 으로 수정

kyukong committed:
docs: 로또 미션에 관한 간략한 기능 설명 추가 및 구현 기능 수정

kyukong committed:
refactor: 더 간결하게 의미를 나타내기 위해 Payment 로 수정

kyukong committed:
refactor: 구입 가능한 로또 개수 반환하는 기능을 Payment 에게 부여

- 구입 가능한 로또 개수는 구입 금액에 속해있는 값이므로 이동
- 역할이 없는 LottoMachine 객체 삭제

kyukong commented:
소니! 궁금한 것이 하나 더 있습니다. 😁

저는 이펙티브 자바에서 toString 에 대해 `디버깅 시에 편리하고 로그를 남길 때 사용한다` 라고 이해했습니다. (아이템 12) 그러므로 매번 정의하라고 하더군요. 하지만 이전 미션의 피드백과 다른 크루들의 피드백들을 보면 toString 을 사용하지 않는 분이 더 많아 보이더라구요. 접근 제어자가 public 이기도해서 당장 기능에 사용하지 않아도 view 의 역할을 대신하기 때문인 것 같습니다.

소니는 평소 toString 을 사용하는지, 사용하지 않는다면 그 이유가 궁금합니다...! 저에겐 아직 경험이 많이 부족해 실무자의 의견이 궁금합니다. 👀

sonypark commented on src/main/java/lotto/domain/Payment.java:
```diff
@@ -0,0 +1,53 @@
+
+    private final int payment;
+
+    public Payment(final String payment) {
```
원시값 포장 👍

sonypark replied:
인자 타입을 String이 아닌 int로 받는건 어떨까요?

sonypark commented on src/main/java/lotto/domain/Payment.java:
```diff
@@ -0,0 +1,53 @@
+        if (isBlank(payment) || !isNumber(payment)) {
+            throw new IllegalArgumentException(ERROR_ONLY_NUMBER);
+        }
+        int number = Integer.parseInt(payment);
```
String으로 받아서 숫자에 대한 검증을 해주고 있군요~ int로 받으면 이 부분의 로직은 분리될 수 있을 것 같은데 어떠신가요?

sonypark commented on src/main/java/lotto/domain/Lottos.java:
```diff
@@ -0,0 +1,38 @@
+        }
+        return randomLottos;
+    }
+}
```
Lottos에서 자기 자신을 만드는 군요.
Lottos를 만들어주는 역할을 하는 클래스를 따로 두는건 어떤가요?

sonypark commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,86 @@
+
+    public Lotto() {
+        this(selectNumbers());
+    }
```
여기도 Lotto 스스로 자신을 만들고 있네요.
이런 경우 테스트 코드 작성하기 어렵지 않으셨나요?

sonypark commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,86 @@
+        return IntStream.range(MINIMUM_NUMBER, MAXIMUM_NUMBER + 1)
+            .boxed()
+            .collect(Collectors.toList());
+    }
```
로또가 생성 될 때마다 매번 이 로직이 동작하고 있는데, 로또번호는 정해져 있어서 재사용이 가능할 것 같은데. 재사용성을 높이는 방법은 없을까요?

sonypark commented on src/main/java/lotto/domain/Ball.java:
```diff
@@ -0,0 +1,50 @@
+    @Override
+    public String toString() {
+        return Integer.toString(number);
+    }
```
이 메서드를 오버라이드 해주신 이유는 무엇인가요?

sonypark commented on src/main/java/lotto/domain/Ball.java:
```diff
@@ -0,0 +1,50 @@
+        Ball ball = (Ball)obj;
+        return this.number == ball.number;
+    }
+}
```
equals를 오버라이드 하셨네요. 아마 숫자 비교를 위해 하셨을 것 같은데. 비교할 수 있는 다른 방법은 없을까요?
추가로 equals와 hashcode를 언제 오버라이드 하는지, 둘 중 하나만 해도 되는지 등에 대해 공부해보시면 좋을 것 같아요🧐

sonypark commented on src/main/java/lotto/domain/Lottos.java:
```diff
@@ -0,0 +1,38 @@
+            boolean isBonus = matchingCount == CHECK_BONUS_COUNT && lotto.contains(winningLotto.getBonusBall());
+            lottoResult.increaseRankCount(Rank.getRank(matchingCount, isBonus));
+        }
+    }
```
이 부분은 당첨번호를 매칭하는 로직인데 Lottos가 갖고 있을 책임이 맞을까요?

sonypark commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -0,0 +1,91 @@
+        try {
+            List<String> numbers = Arrays.stream(inputWinNumber().split(DELIMITER))
+                    .map(String::trim)
+                    .collect(Collectors.toList());
```
이부분은 View에서 담당하는게 더 자연스러울 것 같은데 어떻게 생각하시나요~

sonypark commented on src/main/java/lotto/domain/WinningLotto.java:
```diff
@@ -0,0 +1,28 @@
+    public WinningLotto(Lotto winningLotto, Ball bonusBall) {
+        validateDuplicatedNumber(winningLotto, bonusBall);
+        this.winningLotto = winningLotto;
+        this.bonusBall = bonusBall;
```
null 검증이 필요할 것 같네요

sonypark reviewed (changes requested):
유콩 안녕하세요~~
이번 리뷰를 맡은 소니입니다! 😃
로또 미션하느라 고생 많으셨어요.
지난 미션에 배운 내용을 적용하려고 노력하신 것 같아요. 잘 하셨습니다!
코드 중 개선해볼 수 있는 부분에 대해 의견남겨 보았는데요. 한 번 읽어보시고 생각해보시면 좋을 것 같습니다.

p.s. 제 피드백이 틀릴 수 있으니 무조건 적으로 반영하지 마시고 공부를 통해 납득할 수 있는 경우에 반영해주세요. 또한 더 좋은 방법이 있는 경우에도 알려주시면 좋을 것 같아요. 더 나은 구조를 위한 토론은 언제나 환영입니다.🙌🏻

sonypark commented:
> 소니! 궁금한 것이 하나 더 있습니다. 😁
> 
> 저는 이펙티브 자바에서 toString 에 대해 `디버깅 시에 편리하고 로그를 남길 때 사용한다` 라고 이해했습니다. (아이템 12) 그러므로 매번 정의하라고 하더군요. 하지만 이전 미션의 피드백과 다른 크루들의 피드백들을 보면 toString 을 사용하지 않는 분이 더 많아 보이더라구요. 접근 제어자가 public 이기도해서 당장 기능에 사용하지 않아도 view 의 역할을 대신하기 때문인 것 같습니다.
> 
> 소니는 평소 toString 을 사용하는지, 사용하지 않는다면 그 이유가 궁금합니다...! 저에겐 아직 경험이 많이 부족해 실무자의 의견이 궁금합니다. 👀

네 toString을 재정의하는 이유는 유콩말처럼 로깅과 디버깅을 할 때 가독성을 높이기 위해서 입니다. 재정의 하지 않으면 클래스 이름과 메모리주소만 나오기 때문에 객체 안에 들어있는 값이 무엇인지 알 수 없기 때문인데요. `[Oracle문서: toString에 대한 정의](https://docs.oracle.com/javase/10/docs/api/java/lang/Object.html#toString())`에도 보면 모든 서브클래스에 toString()을 재정의 할 것을 권장하고 있네요. 단, 재정의 할 때는 객체가 가진 정보를 잘 표시해주는게 중요한데 보통 수동으로 하지 않고 IDE의 자동생성 기능을 이용하는걸 권장합니다. 유콩이 Ball에 toString()을 재정의 해주셨는데, IDE로 재정의한것과 어떤 차이가 있는지 한 번 확인해보시면 좋을 것 같아요. 실무에서도 사용하는지 질문하셨는데, 저희 팀 같은 경우는 로깅이 필요한 부분(DTO, 변경이력을 남기는 클래스 등)에 toString() 메서드를 재정의해서 사용하는 경우가 있습니다. 그런데 로깅을 할 때 객체를 통째로 로그에 넣어주기 보다는 필요한 값만 추출해서 넣어주는 경우가 많아서 모든 클래스에 toString을 재정의하지는 않고 있습니다.

kyukong committed:
refactor: 당첨 여부를 비교하는 로직을 LottoResult 에게 부여

- 메소드 이동에 따른 테스트 위치 수정

kyukong committed:
refactor: Payment 생성자 파리미터 타입을 int 로 변경

- 파라미터 타입을 String 에서 int 로 변경하여 생성자 내부에 String 관련 검증 로직을 Input 으로 이동
- PaymentTest 에서 String 관련 검증 테스트 하던 것을 InputTest 로 이동
- 입력 값을 임의로 지정하기 위해 전략 패턴 사용하여 Entering 인터페이스로 분리

kyukong committed:
refactor: Ball 과 Lotto 의 파라미터 타입 int 로 변경

- 타입 변경에 따른 테스트 코드 수정

kyukong committed:
refactor(Lotto): 매번 중복하여 생성하는 로또 총 숫자 리스트 생성 방법 수정

kyukong commented:
### toString 관련
toString 을 재정의 하는 목적이 `디버깅` 과 `로깅` 인 만큼 해당 필요성이 있을 때에만 정의를 해주는 것이 좋겠군요. 이번 미션에서는 로그를 남기지 않기 때문에 로깅을 제외하더라도 디버깅하기 어려운 값도 없을 뿐더러 테스트에서 잘못된 값이 입력될 경우를 확인하므로 굳이 필요하지는 않아 보입니다! Ball 에서 재정의한 toString 도 삭제하였습니다. 😁
(Ball 객체에서 toString 을 재정의한 것은 초반에 모든 객체에 재정의 하는 것이 맞은가를 고민하다가 실수로 객체 하나에만 추가하고 삭제를 안했습니다...!! 🤣)

kyukong committed:
refactor(Ball): toString 오버라이딩 삭제

kyukong committed:
refactor(Ball): equals 메서드 로직 보완 및 hashCode 메서드 추가

- 각 메서드의 테스트 추가

kyukong commented:
### Ball 객체의 equals 관련
> equals를 오버라이드 하셨네요. 아마 숫자 비교를 위해 하셨을 것 같은데. 비교할 수 있는 다른 방법은 없을까요?

equals 를 사용해본 적이 없어 연습겸 추가해봤습니다. 🤣 다른 방법을 생각해 본다면 같은 로직을 가지고 있는 별도의 메소드를 생성하는 것밖에는 생각이 안나네요....😂😂 equals 에 대해 다시 찾아본 결과, equals 는 보통 구현한 프로그램 내에서 `실제로는 다른 주소값을 갖는 객체지만 동일한 객체(인스턴스)로 판단해도 되는 값` 에 한해 구현한다고 이해했습니다. 로또 미션에서 본다면 로또의 숫자인 Ball 은 Lotto 를 생성할 때마다 각각 생성되는 별도의 객체지만 실질적 객체 비교는 무의미하고 객체 내 숫자의 동일 여부가 더 중요하다는 생각이 들어 equals 를 삭제하지 않았습니다. 로직에서 타입에 관한 검증이 부족해 추가하였습니다!

> 추가로 equals와 hashcode를 언제 오버라이드 하는지, 둘 중 하나만 해도 되는지 등에 대해 공부해보시면 좋을 것 같아요🧐

이펙티브 자바에 의하면, equals 를 재정의하여 프로그램 내에서 동등하다고 판단하였으나 인스턴스를 HashMap 또는 HashSet 같은 컬렉션의 원소로 사용할 때 hashCode 의 값을 기준으로 정의하므로 사용 시 문제가 발생할 수 있다고 합니다. equals 와 동일한 맥락으로 값이 동일한 것은 동등한 객체로 판단하였기 때문에 해시값도 일치해야 하여 필수로 재정의 해야 한다고 이해했습니다. 😁

kyukong committed:
refactor: 지난 주 당첨번호 입력 값 변환 과정 Input 으로 이동

kyukong committed:
fix(Lotto): Ball 의 toString을 제거하여 getter 로 대체

kyukong committed:
fix(Ball): number getter 추가

kyukong committed:
fix(Input): 지난 주 당첨번호 입력 시 양끝 공백 삭제 후 검증

kyukong committed:
refactor: 각 성격에 맞는 상수 위치로 이동

kyukong committed:
refactor: WinningLotto 생성자 파라미터 null 처리 추가

- null 처리 및 중복 관련 테스트 추가
- 객체의 이름 WinningLotto 와 인스턴스 변수인 지난 주 당첨번호의 이름 winningLotto 를 구분하기 위해 인스턴스 변수명 수정

kyukong committed:
refactor: private 생성자로 인스턴스를 만들지 않아야 하는 객체 생성 차단

kyukong committed:
refactor: 재할당 금지 및 외부에서 값을 변경할 수 없도록 수정

kyukong committed:
refactor: Lotto 와 Lottos 의 생성자 축소

- 해당 생성자를 사용하지 않아도 구현할 수 있기에 삭제

kyukong committed:
refactor: 생성자 호출 관련 메소드 별도 클래스 생성하여 이동

kyukong commented:
### 생성자를 호출하는 클래스(Creator) 관련
Controller 가 수행하는 역할이 많아 보여 로직을 줄이고 싶어도 어떤 것을 줄일지 고민이 많았는데 피드백을 보고 생성자 호출 관련 로직만을 분리해보았습니다...! 통일성을 위해 생성자 호출 외에 다른 로직이 없는 것들도 별도의 메소드를 만들어 호출하였습니다. 입력값에 대해 생성자를 호출하여 반환하므로 view 로 분류해보았는데 괜찮을까요?
(Lotto 와 Lottos 의 자기 자신을 생성하여 생성자를 호출 하던 것은 다른 방식으로 구현할 수 있을듯해 생성자를 삭제하였습니다!)

그리고 피드백에서 주신 자기 자신의 객체 정보를 생성하여 생성자를 호출하는 형태가 왜 좋지 않은지 모르겠습니다. 피드백에서 언급해주신 테스트를 진행하기 어려워서일까요?🥺 어차피 또 다른 생성자를 호출하는 형태므로 마지막으로 호출하는 생성자만을 테스트하면 간접적으로 테스트 할 수 있지 않을까요? 아니면 자기 자신의 정보를 자기가 생성한 다는 것이 의미가 맞지 않아서 일까요?

kyukong committed:
chore(Ball): 주석 삭제

kyukong committed:
refactor(LottoResult): 메소드 위치 조정

sonypark commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,83 @@
+    private static final String ERROR_BALL_COUNT = BALL_COUNT + "개의 숫자를 입력해주세요";
+    private static final String ERROR_DUPLICATED_NUMBER = "번호가 중복됩니다!";
+
+    private final List<Ball> lotto = new ArrayList<>();
```
아래 두 필드 모두 필요한걸까요?
- List<Integer> lottoTotalNumbers = new ArrayList<>()
- List<Ball> lotto = new ArrayList<>();

sonypark commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,83 @@
+                .boxed()
+                .collect(Collectors.toList()));
+    }
+
```
로또번호를 캐싱해주고 있네요.
캐싱을 해주는 건 좋은데 캐싱을 해주는 위치가 Lotto에 있는게 좋을까요?
1번-45번까지 번호를 갖는건 Lotto와 Ball 중 어떤 속성인가요?

sonypark commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,83 @@
+        for (Integer number : numbers) {
+            this.lotto.add(new Ball(number));
+        }
+    }
```
Ball이라는 VO를 만들었는데 Integer로 받고 있네요.
VO를 활용해볼까요?

sonypark commented on src/main/java/lotto/domain/Ball.java:
```diff
@@ -0,0 +1,43 @@
+package lotto.domain;
+
+import java.util.Objects;
+
+public class Ball {
+    public static final int MINIMUM_NUMBER = 1;
+    public static final int MAXIMUM_NUMBER = 45;
+    private static final String ERROR_LOTTO_NUMBER = "로또 숫자가 아닙니다";
+
+    private final int number;
+
+    public Ball(final int number) {
+        validateLottoNumber(number);
+        this.number = number;
+    }
```
로또 번호를 VO로 만드셨군요!
VO로 만든 이유가 있으신가요?

sonypark commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,83 @@
+        return selectedIntNumbers.stream()
+            .sorted()
+            .collect(Collectors.toList());
+    }
```
Lotto에서 랜덤으로 로또 번호를 생성하고 있는데요.
Lotto 클래스의 역할이 너무 많은 것 같은데 역할에 맞는 클래스로 나누어 책임을 분리해보는건 어떤가요?

sonypark commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,83 @@
+        return (int) lotto.stream()
+                .filter(otherLotto::contains)
+                .count();
+    }
```
compareTo 메서드를 오버라이드 하셨는데, compareTo는 언제 오버라이드를 할까요?
현재 로직은 두 로또가 가진 번호 중 같은 번호의 갯수를 리턴하는 로직인데 compareTo 메서드의 목적에 부합할까요?

sonypark commented on src/main/java/lotto/domain/Payment.java:
```diff
@@ -0,0 +1,38 @@
+        if ((number % Lotto.LOTTO_PRICE) != 0) {
+            throw new IllegalArgumentException(ERROR_LOTTO_PRICE);
+        }
+    }
```
BuyLotto라는 메서드명이 조금 어색한데 변경해보는건 어떨까요?

sonypark commented on src/main/java/lotto/domain/LottoResult.java:
```diff
@@ -0,0 +1,41 @@
+    public void match(final Lottos lottos, final WinningLotto winningLotto) {
+        for (Lotto lotto : lottos.getLottos()) {
+            int matchingCount = lotto.compareTo(winningLotto.getWinningNumbers());
+            boolean bonus = matchingCount == CHECK_BONUS_COUNT && lotto.contains(winningLotto.getBonusBall());
```
당첨번호를 매칭하는 로직은 WinningNumbers에게 위임해도 될 것 같은데 어떠세요?

sonypark commented on src/main/java/lotto/domain/Lottos.java:
```diff
@@ -0,0 +1,25 @@
+
+    public List<Lotto> getLottos() {
+        return new ArrayList<>(lottos);
+    }
```
외부에서 lottos.add(lotto)를 하면 어떻게 되나요?

sonypark commented on src/main/java/lotto/view/Creator.java:
```diff
@@ -0,0 +1,65 @@
+            lottos.add(new Lotto(Lotto.selectRandomNumbers()));
+        }
+        return new Lottos(lottos);
+    }
```
Lottos를 생성하는 로직은 중요한 비즈니스 로직인데요.
다음과 같은 문제가 있습니다.
- 테스트가 없다.
- 해당 책임을 같는 클래스에 있지 않다(Creator의 역할은 무엇인가?)
- 로또 생성 로직이 바뀌었을 때 변경에 유연하지 못하다. (자동 생성-> 수동 생성)

Lottos를 생성하는 클래스를 만들어보는건 어떤가요?

sonypark commented on src/main/java/lotto/domain/Profit.java:
```diff
@@ -0,0 +1,12 @@
+        profitRate = (double) totalMoney / money;
+        return profitRate;
+    }
+}
```
이 클래스가 필요한지는 잘 모르겠네요😅
LottoResult에서 담당해줄 수 있지 않을까요?

sonypark reviewed (changes requested):
유콩 안녕하세요 :)
작성해주신 코드 잘 봤습니다😃
2단계로 넘어가기 전에 객체의 역할/책임을 분리하는 것에 초점을 두고 리팩토링을 해보면 좋을 것 같습니다!
(아 그리고 테스트가 깨지는게 몇개 보이던데 확인 부탁드려요🙏🏻)

우선 전반적인 코멘트를 남겨봅니다.

1. 생성자를 호출하는 클래스(Creator)를 만들어 주셨는데요. 이 클래스의 역할은 View에서 입력값을 받아 도메인 객체를 생성하고 있는데 view 패키지에 들어있네요. view 패키지에 있는게 맞을까요? 역할은 컨트롤러의 역할을 하고 있는 것 같아서요.

2. 자기 자신의 객체 정보를 생성하는 생성자가 좋지 않은 이유는 생성자의 역할을 생각해보시면 될 것 같습니다. 정적 팩토리 메서드를 사용해서 객체를 생성하거나 복수의 생성자가 있어 내부에서 생성자를 호출하는 경우가 아니라면 클래스 내부에서 생성자를 호출할 일이 있을까요? 로또 하나는 6개의 로또 번호를 갖고 있습니다. 그렇다면 로또가 생성될 때 6개의 로또 번호를 받아야 할거에요. (내부에서 스스로 생성하는게 아니라) 따라서 객체의 역할과 책임의 분리가 필요한 것 같아 피드백을 드렸습니다. 관련 코멘트는 아래에 남겼으니 한 번 고민해보시면 좋을 것 같아요~ 이해 안되시거나 궁금한 점 있으면 언제든 코멘트 남겨주세요~

kyukong committed:
fix: null 체크 관련 테스트 실패하여 수정

kyukong committed:
refactor: Creator 클래스 controller 로 이동

- view 에서 값을 받아와 domain 객체를 생성하므로 controller 로 이동함

kyukong committed:
refactor: Lotto 생성자 파라미터 List<Ball> 로 변경

- 그에 따른 코드 및 테스트 코드 수정

kyukong committed:
refactor: Lottos getter 수정 불가능 및 Lotto 매칭 메소드명 수정

kyukong committed:
refactor: 수익률 계산 역할을 LottoResult 에게 부여

kyukong committed:
refactor: 로또 생성 역할을 하는 LottoGenerator 객체 생성

kyukong committed:
refactor: 당첨번호 매칭 역할 WinningLotto 에게 부여

sonypark reviewed (approved):
피드백은 2단계에서 이어서 진행하겠습니다 :)
